### ISSUE TITLE
Can I instantiate an alias-recursive generic type?

### ISSUE THREAD

**Issue by sh01ch1** [crystal-lang/crystal#3031]
``` crystal
class A
  alias Alias = Array(Alias)
end

A::Alias.new
```

prints:

```
Error in line 5: undefined method 'new' for A::Alias:Class
```

I let Crystal be quiet like this:

``` crystal
class A
  alias Alias = Array(Alias)
  alias Alias2 = Alias
end

A::Alias2.new # OK
```

Thank you.

**Comment by asterite** [comment]
No, this should eventually give a compile error. A recursive alias must have a non-recursive base type.

**Comment by sh01ch1** [comment]
Thanks! So... instead of alias, is the following class definition permitted in Crystal? It seems working.

```crystal
class A < Array(A)
end

A.new
```

**Comment by asterite** [comment]
Yes, it works, though I wouldn't advise using it right now as there are some bugs related to generic types and inheritance.

What do you need to do?

**Comment by sh01ch1** [comment]
Thank you! For example:

```crystal
require "json"

class Model
  class Portable < Hash(String, String | Array(Portable) | Nil)
  end

  getter children = [] of Model
  property attribute : String?

  def to_portable
    h = Portable.new
    h["attribute"] = attribute
    h["children"] = children.map{|i| i.to_portable.as(Portable)} # this 'as' is needed for type inference
    h
  end

  def to_json
    to_portable.to_pretty_json
  end
end

m = Model.new
m.attribute = ":)"
m.children << Model.new
puts m.to_json
```

output:

```
{
  "attribute": ":)",
  "children": [
    {
      "attribute": null,
      "children": []
    }
  ]
}
```

Actually, my real use case is more and more complex.

**Comment by danielpclark** [comment]
I need this feature.  I need to have a NamedTuple where it's values can be either String or another NamedTuple of the same kind.  So a recursive named tuple.

**Comment by vladfaust** [comment]
@asterite 

```crystal
class A < Array(A)
end

A.new
```

Doesn't work anymore...

**Comment by asterite** [comment]
Closed in favor of #3804 (kinda duplicate)
